### PR TITLE
[SDK] ochestrator

### PR DESCRIPTION
**Pull request by brynn-code**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 1]
```diff
@@ -383,7 +383,7 @@ def _get_data_path(self, run: Union[str, Run]) -> str:
         """Get the outputs file path of the run."""
         local_storage = self._get_local_storage(run)
         # TODO: what if the data is deleted?
-        if not local_storage._data_path or not os.path.exists(local_storage._data_path):
+        if local_storage._data_path and not os.path.exists(local_storage._data_path):
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:386]: Discuss about this.

File: src/promptflow/promptflow/_cli/_pf/_experiment.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        add_params=[add_param_name] + base_params,
+        subparsers=subparsers,
+        help_message="Start an experiment.",
+        action_param_name="sub_action",
+    )
+
+
+def add_experiment_parser(subparsers):
+    # TODO: Update the help after the command is finalized.
+    experiment_parser = subparsers.add_parser(
```
Comment (brynn-code) [inline src/promptflow/promptflow/_cli/_pf/_experiment.py:123]: Note: experiment is suppressed for now.

File: src/promptflow/promptflow/_sdk/entities/_run.py [hunk 3]
```diff
@@ -150,6 +152,7 @@ def __init__(
         # init here to make sure those fields initialized in all branches.
         self.flow = flow
         self._use_remote_flow = is_remote_uri(flow)
+        # What's this?
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/entities/_run.py:155]: Sync about the existing experiment name.

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        super().__init__(**kwargs)
+
+
+class ExperimentRunSubmitter(RunSubmitter):
+    @classmethod
+    def _validate_inputs(cls, run: Run):
+        # Do not validate run/data field, as we will resolve them in _resolve_input_dirs.
+        return
+
+    def _resolve_input_dirs(self, run: ExperimentRun):
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:105]: Support inputs.

File: src/promptflow/promptflow/_cli/_pf/entry.py [hunk 5]
```diff
@@ -97,6 +100,7 @@ def get_parser_args(argv):
     add_run_parser(subparsers)
     add_config_parser(subparsers)
     add_tool_parser(subparsers)
+    add_experiment_parser(subparsers)
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/entry.py:103]: hide the experiment cli in an pf.config

File: src/promptflow/promptflow/_sdk/_constants.py [hunk 6]
```diff
@@ -394,3 +397,14 @@ class DownloadedRun:
...
+
+
+class ExperimentNodeType(object):
+    FLOW = "flow"
+    PYTHON = "python"
+
+
+class ExperimentStatus(object):
+    NOT_STARTED = "NotStarted"
+    RUNNING = "Running"
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_constants.py:409]: InProgress?

File: src/promptflow/promptflow/_sdk/_orm/experiment.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    created_on = Column(TEXT, nullable=False)  # ISO8601("YYYY-MM-DD HH:MM:SS.SSS"), string
+    status = Column(TEXT, nullable=False)
+    description = Column(TEXT)  # updated by users
+    properties = Column(TEXT)
+    archived = Column(Boolean, default=False)
+    nodes = Column(TEXT)  # json(list of json) string
+    node_runs = Column(TEXT)  # json(list of json) string
+    # NOTE: please always add columns to the tail, so that we can easily handle schema changes;
+    #       also don't forget to update `__pf_schema_version__` when you change the schema
+    # NOTE: keep in mind that we need to well handle runs with legacy schema;
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_orm/experiment.py:35]: @lalala123123 Please take a look

File: src/promptflow/promptflow/_sdk/_orm/session.py [hunk 8]
```diff
@@ -77,12 +79,14 @@ def mgmt_db_session() -> Session:
...
-        from promptflow._sdk._orm import Connection, RunInfo
+        from promptflow._sdk._orm import Connection, Experiment, RunInfo
 
         create_or_update_table(engine, orm_class=RunInfo, tablename=RUN_INFO_TABLENAME)
+        create_or_update_table(engine, orm_class=Experiment, tablename=EXPERIMENT_TABLE_NAME)
         create_table_if_not_exists(engine, CONNECTION_TABLE_NAME, Connection)
 
-        create_index_for_run_if_not_exists(engine)
+        create_index_if_not_exists(engine, RUN_INFO_CREATED_ON_INDEX_NAME, RUN_INFO_TABLENAME, "created_on")
+        create_index_if_not_exists(engine, EXPERIMENT_CREATED_ON_INDEX_NAME, EXPERIMENT_TABLE_NAME, "created_on")
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_orm/session.py:89]: only create if enabled private features?

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 9]
```diff
@@ -206,6 +208,11 @@ def flows(self) -> FlowOperations:
         """Operations on the flow that can manage flows."""
         return self._flows
 
+    @property
+    def experiments(self) -> ExperimentOperations:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_pf_client.py:212]: shall we hide or add some experimental notice here?

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+            if not action:
+                raise UserErrorException(f"Experiment has circular dependency {edges!r}")
+
+        logger.debug(f"Experiment nodes resolved order: {[node.name for node in resolved_nodes]}")
+        return resolved_nodes
+
+    def _run_node(self, node, experiment, data_dict, run_dict) -> Run:
+        if node.type == ExperimentNodeType.FLOW:
+            return self._run_flow_node(node, experiment, data_dict, run_dict)
+        elif node.type == ExperimentNodeType.PYTHON:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:101]: Is this code type?

File: src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+            config=Configuration(overrides={Configuration.RUN_OUTPUT_PATH: run_output_path}),
+        )
+        logger.debug(f"Creating run {run.name}")
+        return self.run_submitter.submit(run)
+
+    def _run_script_node(self, node, experiment):
+        pass
+
+
+class ExperimentRun(Run):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py:129]: what's a experiemnt Run? add some docstring

File: src/promptflow/promptflow/_cli/_pf/_experiment.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        add_params=[add_param_name] + base_params,
+        subparsers=subparsers,
+        help_message="Start an experiment.",
+        action_param_name="sub_action",
+    )
+
+
+def add_experiment_parser(subparsers):
+    experiment_parser = subparsers.add_parser(
+        "experiment", description="A CLI tool to manage experiment for prompt flow.", help="pf experiment"
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/_experiment.py:123]: mention this is [experimental] and will change.

File: src/promptflow/promptflow/_sdk/operations/_experiment_operations.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        self,
+        max_results: Optional[int] = MAX_RUN_LIST_RESULTS,
+        *,
+        list_view_type: ListViewType = ListViewType.ACTIVE_ONLY,
+    ) -> List[Experiment]:
+        """List experiments.
+
+        :param max_results: Max number of results to return. Default: MAX_RUN_LIST_RESULTS.
+        :type max_results: Optional[int]
+        :param list_view_type: View type for including/excluding (for example) archived runs. Default: ACTIVE_ONLY.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_experiment_operations.py:35]: "run"  -> experiments

File: src/promptflow/promptflow/_sdk/operations/_experiment_operations.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        orm_experiments = ORMExperiment.list(max_results=max_results, list_view_type=list_view_type)
+        return safe_parse_object_list(
+            obj_list=orm_experiments,
+            parser=Experiment._from_orm_object,
+            message_generator=lambda x: f"Error parsing experiment {x.name!r}, skipped.",
+        )
+
+    @monitor_operation(activity_name="pf.experiment.get", activity_type=ActivityType.PUBLICAPI)
+    def get(self, name: str) -> Experiment:
+        """Get a run entity.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_experiment_operations.py:49]: "run" -> experiment

File: src/promptflow/promptflow/_sdk/schemas/_experiment.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    source = NestedField(ToolSourceSchema, required=True)
+    inputs = fields.Dict(keys=fields.Str)
+    # runtime field, only available for cloud run
+    runtime = fields.Str()  # TODO: Revisit the required fields
+    display_name = fields.Str()
+    environment_variables = fields.Dict(keys=fields.Str, values=fields.Str)
+
+
+class FlowNodeSchema(RunSchema):
+    class Meta:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/schemas/_experiment.py:31]: not enherit RunSchema? fully rewrite?

**Comment by zhengfeiwang** [comment]
Nit. the PR title should be o**r**chestrator

**Review by zhengfeiwang — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Pull request merged**